enhancements/automated-service-ca-rotation.md

```diff
@@ -0,0 +1,204 @@
+- Implement library-go support for [hitless
+rotation](https://diogomonica.com/2017/01/11/hitless-tls-certificate-rotation-in-go/)
+so that consumers of serving certs and CA bundles can trivially support automated
+refresh.
+
+## Proposal
+
+- Detect when the current CA certificate has passed 1/2 of its validity bounds
+  (i.e. 6 months has passed of a 1 year validity) and generate a new CA.
+  - The new CA should have the same `Subject.CommonName` as the current CA to
+    ensure that trust chaining is possible.
+  - This provides a 6 month grace period for consumers of serving certs and ca
+    bundles to be refreshed.
+
+- Include both the new and old CAs' primary certificate in a new CA bundle.
+  - The previous CA's primary certificate will serve to bridge trust between an
+    unrefreshed server (serving with a cert generated from the previous CA) and
+    a refreshed client (validating trust with the new CA bundle). The existing
+    CA's primary certificate will ensure that certificates generated from the
+    new CA can be trusted.
+
+- Generate an intermediate CA certificate with the same public key as the new CA
+  but signed with the private key of the current CA.
+  - This intermediate certificate should be included with newly-generated serving
+    certificates, and serves to bridge trust between the CA bundle of an
+    unrefreshed client (validating trust with the old CA bundle) and a service
+    serving with a cert generated from the new CA.
+
+- Compare the `AuthorityKeyId` of a serving cert with the `SubjectKeyId` of the
+  current CA to determine whether the serving cert was issued by the current CA.
+  - If the serving cert is determined not to have been issued by the current CA,
+    it needs to be regenerated.
+  - Previously, `Subject.CommonName` of a serving cert was compared with
+    `Issuer.CommonName` of the CA. Since ensuring trust between old and new CAs
+    requires that `Subject.CommonName` be reused, this comparison will no longer
+    be sufficient to determine lineage.
+  - This requires that all new certificates - whether CA or entity - have
+    `AuthorityKeyId` and `SubjectKeyId` set.
+
+- Provide an indication that rotation has taken place in the status of the
+  operator's CR.
+
+### Risks and Mitigations
+
+Automatic rotation is only effective if consumers of certs and CA bundles will
+also refresh automatically.
+
+The proposed rotation scheme does not extend the useful lifespan of serving certs
+and CA bundles from the pre-rotation CA. Their lifespan will still be 1
+year. Instead, the rotation schema provides a 6 month window when both the old
+and new CAs will be trusted. Within that 6 month window, it is critical that
+services that consume serving certs and clients that consume CA bundles start
+using the serving certs and CA bundle provided by the new CA. For some consumers
+this may happen automatically (e.g. via restart on change or hitless
+rotation. Others may need to be manually restarted.
```

Comment:
can you write down the exact dates (worst case) ? When did we release 4.1 (when the first public beta)? When could the first CA been created and when would it expire? Until when do customers have to update?

Reply:
Done